stellargraph/connector/neo4j/graph.py

```diff
@@ -58,25 +66,94 @@ def nodes(self):
         result = self.graph_db.run(node_ids_query)
         return [row["node_id"] for row in result.data()]
 
-    def node_features(self, node_ids):
+    def cache_nodes(self, dtype="float32"):
+        nodes = self.nodes()
+        features = self._node_features_from_db(nodes)
+        self._nodes = convert.convert_nodes(
+            pd.DataFrame(features, index=nodes),
+            name="nodes",
+            default_type=self._node_type,
+            dtype=dtype,
+        )
+
+        # cache feature size too
+        self._node_feature_size = self._nodes.features_of_type(self._node_type).shape[1]
+
+    def _node_features_from_db(self, nodes):
+        if nodes is None:
+            where_clause = ""
+            valid = slice(None)
+        else:
+            where_clause = " WHERE node.ID = node_id"
+            if isinstance(nodes, np.ndarray):
+                valid = nodes != None
+                # we need to create a list of integers to run the neo4j query with
+                nodes = [int(node) for node in nodes]
+            elif isinstance(nodes, list):
+                valid = np.array(nodes) != None
+            else:
+                # single node
+                valid = np.array([nodes is not None])
+                nodes = [nodes]
+
+        # None's should be filled with zeros in the feature matrix
+        features = np.zeros((len(nodes), self.node_feature_sizes()[self._node_type]))
+
+        # fill valid locs with features
+        feature_query = f"""
+            UNWIND $node_id_list AS node_id
+            MATCH(node) {where_clause}
+            RETURN node.features as features
+            """
+        result = self.graph_db.run(feature_query, parameters={"node_id_list": nodes})
+        rows = result.data()
+
+        # this method currently doesn't handle any other invalid IDs. If there are other
+        # invalid ids, we should raise an error
+        if len(rows) != sum(valid):
+            raise ValueError(
+                "nodes: Found values that did not return any results from the database."
```

Comment:
I wonder how easy it is for us to be specific about the values that we're missing?

Reply:
Hmm yeah can do something like
```
WITH {{ID: node.ID, features: node.features}} AS node_data
RETURN node_data
```
then check the IDs being returned